oidc/client_race_test.go

```diff
@@ -0,0 +1,61 @@
+// This file contains tests which depend on the race detector being enabled.
+// +build race
+
+package oidc
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+)
+
+type testProvider struct {
+	baseURL string
+}
+
+func (p *testProvider) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Path != discoveryConfigPath {
+		http.NotFound(w, r)
+		return
+	}
+
+	cfg := ProviderConfig{
+		Issuer: p.baseURL,
+	}
+	json.NewEncoder(w).Encode(&cfg)
+}
+
+// If this test fails by triggering the race detector, not by calling t.Error or t.Fatal.
+func TestProviderSyncRace(t *testing.T) {
+	prov := &testProvider{}
+
+	s := httptest.NewServer(prov)
+	defer s.Close()
+	prov.baseURL = s.URL
+
+	cliCfg := ClientConfig{
+		HTTPClient:     http.DefaultClient,
+		ProviderConfig: ProviderConfig{Issuer: s.URL},
+	}
+	cli, err := NewClient(cliCfg)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	// SyncProviderConfig beings a goroutine which writes to the client's provider config.
+	c := cli.SyncProviderConfig(s.URL)
+	defer func() {
+		// stop the background process
+		c <- struct{}{}
+	}()
+
+	// Iterate serveral times to increase the likelyhood of triggering the race detector.
+	for i := 0; i < 100; i++ {
+		time.Sleep(5 * time.Millisecond)
```

Comment:
The next sync will happen in `MinimumProviderConfigSyncInterval` which is 1 min. maybe we make that tunable for test ?